models/issue_tracked_time.go

```diff
@@ -74,30 +132,63 @@ func (opts *FindTrackedTimesOptions) ToSession(e Engine) *xorm.Session {
 	return x.Where(opts.ToCond())
 }
 
-// GetTrackedTimes returns all tracked times that fit to the given options.
-func GetTrackedTimes(options FindTrackedTimesOptions) (trackedTimes []*TrackedTime, err error) {
-	err = options.ToSession(x).Find(&trackedTimes)
+func getTrackedTimes(e Engine, options FindTrackedTimesOptions) (trackedTimes TrackedTimeList, err error) {
+	err = options.ToSession(e).Find(&trackedTimes)
 	return
 }
 
+// GetTrackedTimes returns all tracked times that fit to the given options.
+func GetTrackedTimes(opts FindTrackedTimesOptions) (TrackedTimeList, error) {
+	return getTrackedTimes(x, opts)
+}
+
+func getTrackedSeconds(e Engine, opts FindTrackedTimesOptions) (trackedSeconds int64, err error) {
+	return opts.ToSession(e).SumInt(&TrackedTime{}, "time")
+}
+
+// GetTrackedSeconds return sum of seconds
+func GetTrackedSeconds(opts FindTrackedTimesOptions) (int64, error) {
+	return getTrackedSeconds(x, opts)
+}
+
 // AddTime will add the given time (in seconds) to the issue
-func AddTime(user *User, issue *Issue, time int64) (*TrackedTime, error) {
+func AddTime(user *User, issue *Issue, amount int64, created time.Time) (*TrackedTime, error) {
+	sess := x.NewSession()
+	defer sess.Close()
+
+	if err := sess.Begin(); err != nil {
+		return nil, err
+	}
+
+	t, err := addTime(x, user, issue, amount, created)
+	if err != nil {
+		return nil, err
+	}
+	return t, sess.Commit()
+}
+
+func addTime(e Engine, user *User, issue *Issue, amount int64, created time.Time) (*TrackedTime, error) {
+	if created.IsZero() {
+		created = time.Now()
+	}
 	tt := &TrackedTime{
 		IssueID: issue.ID,
 		UserID:  user.ID,
-		Time:    time,
+		Time:    amount,
+		Created: created,
 	}
-	if _, err := x.Insert(tt); err != nil {
+	if _, err := e.Insert(tt); err != nil {
 		return nil, err
 	}
-	if err := issue.loadRepo(x); err != nil {
+	if err := issue.loadRepo(e); err != nil {
 		return nil, err
 	}
+
 	if _, err := CreateComment(&CreateCommentOptions{
```

Comment:
So we should also use `createComment(e,...)`

Reply:
@lunny done

Reply:
(I moved this part of the code from addTime to AddTime